mobile/lib/utils/file_download_util.dart

```diff
@@ -189,50 +190,54 @@ Future<void> downloadToGallery(EnteFile file) async {
     final bool downloadLivePhotoOnDroid =
         type == FileType.livePhoto && Platform.isAndroid;
     AssetEntity? savedAsset;
-    final File? fileToSave = await getFile(file);
-    //Disabling notifications for assets changing to insert the file into
-    //files db before triggering a sync.
-    await PhotoManager.stopChangeNotify();
-    if (type == FileType.image) {
-      savedAsset = await PhotoManager.editor
-          .saveImageWithPath(fileToSave!.path, title: file.title!);
-    } else if (type == FileType.video) {
-      savedAsset =
-          await PhotoManager.editor.saveVideo(fileToSave!, title: file.title!);
-    } else if (type == FileType.livePhoto) {
-      final File? liveVideoFile =
-          await getFileFromServer(file, liveVideo: true);
-      if (liveVideoFile == null) {
-        throw AssertionError("Live video can not be null");
+    // We use a lock to prevent synchronisation to occur while it is downloading
+    // as this introduces wrong entry in FilesDB due to race condition
+    // This is a fix for https://github.com/ente-io/ente/issues/4296
+    await LocalSyncService.instance.getLock().synchronized(() async {
+      final File? fileToSave = await getFile(file);
```

Comment:
We can move this outside. This will fetch the file from remote, and decrypt it.

Reply:
Ack. May be we can use different lock to restrict one download. This will reduce the potential time period during which sync might be blocked while files are getting downloaded.
```dart
await downloadLock.synchronized(() async {
final File? fileToSave = await getFile(file);
});
```

PS: We can keep the previous change to surface dialog progress on UI.

Reply:
I tried multiple other approaches, including the one you suggested and at the end the following was the simplest and with the best results (to my opinion) : 
https://github.com/ente-io/ente/pull/4369/commits/32ed84f48da01d706f59fc70b88cadd3b9d144c9
- moving getFile outside the lock as you initially suggested
- created simple queue for downloads (max 5 concurrent downloads)